modules/core/src/main/java/org/apache/ignite/internal/processors/query/schema/management/SchemaManager.java

```diff
@@ -0,0 +1,1349 @@
+
+        try {
+            for (TableDescriptor tbl : tablesForCache(cacheName)) {
+                tbl.markIndexRebuildInProgress(mark);
+
+                if (mark)
+                    lsnr.onIndexRebuildStarted(tbl.type().schemaName(), tbl.type().tableName());
+                else
+                    lsnr.onIndexRebuildFinished(tbl.type().schemaName(), tbl.type().tableName());
+            }
+        }
+        finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Get table descriptor.
+     *
+     * @param schemaName Schema name.
+     * @param cacheName Cache name.
+     * @param type Type name.
+     * @return Descriptor.
+     */
+    @Nullable public GridQueryTypeDescriptor typeDescriptorForType(String schemaName, String cacheName, String type) {
+        lock.readLock().lock();
+
+        try {
+            SchemaDescriptor schema = schema(schemaName);
+
+            if (schema == null)
+                return null;
+
+            TableDescriptor tbl = schema.tableByTypeName(cacheName, type);
+
+            return tbl == null ? null : tbl.type();
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Gets collection of table for given schema name.
+     *
+     * @param cacheName Cache name.
+     * @return Collection of table descriptors.
+     */
+    public Collection<TableDescriptor> tablesForCache(String cacheName) {
+        lock.readLock().lock();
+
+        try {
+            SchemaDescriptor schema = schema(schemaName(cacheName));
+
+            if (schema == null)
+                return Collections.emptySet();
+
+            List<TableDescriptor> tbls = new ArrayList<>();
+
+            for (TableDescriptor tbl : schema.tables()) {
+                if (F.eq(tbl.cacheInfo().name(), cacheName))
+                    tbls.add(tbl);
+            }
+
+            return tbls;
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Find registered cache info by it's name.
+     */
+    @Nullable public GridCacheContextInfo<?, ?> cacheInfo(String cacheName) {
+        lock.readLock().lock();
+
+        try {
+            SchemaDescriptor schema = schema(schemaName(cacheName));
+
+            if (schema == null)
+                return null;
+
+            for (TableDescriptor tbl : schema.tables()) {
+                if (F.eq(tbl.cacheInfo().name(), cacheName))
+                    return tbl.cacheInfo();
+            }
+
+            return null;
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Find table by it's identifier.
+     *
+     * @param schemaName Schema name.
+     * @param tblName Table name.
+     * @return Table descriptor or {@code null} if none found.
+     */
+    @Nullable public TableDescriptor table(String schemaName, String tblName) {
+        return id2tbl.get(new T2<>(schemaName, tblName));
```

Comment:
Should we get fro map under lock.readLock()?
Looks like we call listeners and update indexes under writeLock on table creation.
There seems to be a race between the add the table to the map and the invocation of listeners.

Reply:
+1, But extra attention is needed. It is necessary to check whether the current thread holds the write lock or not.
Sometimes this method is called under the write lock.

Reply:
Fixed